src/Generator.h

```diff
@@ -640,30 +1567,275 @@ template <class T> class Generator : public Internal::GeneratorBase {
+        generator->call_schedule();
+    }
+
+    // Overloads for first output
+    operator Func() const { 
+        return get_first_output(); 
+    }
+    
+    template <typename... Args> 
+    FuncRef operator()(Args&&... args) const { 
+        return get_first_output()(std::forward<Args>(args)...); 
+    }
+
+    template <typename ExprOrVar> 
+    FuncRef operator()(std::vector<ExprOrVar> args) const { 
+        return get_first_output()(args); 
+    }
+
+    Realization realize(std::vector<int32_t> sizes) { 
+        check_scheduled("realize");
+        return get_first_output().realize(sizes, get_target()); 
+    }
+    
+    template <typename... Args> 
+    Realization realize(Args&&... args) { 
+        check_scheduled("realize");
+        return get_first_output().realize(std::forward<Args>(args)..., get_target()); 
+    }
+
+    template<typename Dst> 
+    void realize(Dst dst) { 
+        check_scheduled("realize");
+        get_first_output().realize(dst, get_target()); 
+    }
+
+    virtual ~GeneratorStub() {}
+
+protected:
+    typedef std::function<std::unique_ptr<GeneratorBase>(const std::map<std::string, std::string>&)> GeneratorFactory;
+
+    template <typename... Args>
+    GeneratorStub(const GeneratorContext &context,
+            GeneratorFactory generator_factory,
+            const std::map<std::string, std::string> &generator_params,
+            const std::vector<std::vector<Internal::FuncOrExpr>> &inputs) {
+        generator = generator_factory(generator_params);
+        generator->target.set(context.get_target());
+        generator->set_inputs(inputs);
+        generator->call_generate();
+    }
+
+    // Output(s)
+    // TODO: identify vars used
+    Func get_output(const std::string &n) { 
+        return generator->get_output(n); 
+    }
+
+    std::vector<Func> get_output_vector(const std::string &n) { 
+        return generator->get_output_vector(n); 
+    }
+
+    bool has_generator() const { 
+        return generator != nullptr; 
+    }
+
+    template<typename Ratio>
+    static double ratio_to_double() {
+        return (double)Ratio::num / (double)Ratio::den;
+    }
+
+private:
+    std::shared_ptr<GeneratorBase> generator;
+
+    Func get_first_output() const { 
+        return generator->get_first_output(); 
+    }
+    void check_scheduled(const char* m) const { 
+        user_assert(generator->schedule_called) << "Must call schedule() before calling " << m << "()"; 
+    }
+
+    explicit GeneratorStub(const GeneratorStub &) = delete;
+    void operator=(const GeneratorStub &) = delete;
+};
+
+}  // namespace Internal
+
+
 }  // namespace Halide
 
+// Use a little variadic macro hacking to allow two or three arguments.
+// This is suboptimal, but allows us more flexibility to mutate registration in
+// the future with less impact on existing code.
+#define _HALIDE_REGISTER_GENERATOR2(GEN_CLASS_NAME, GEN_REGISTRY_NAME) \
```

Comment:
Had to google `__COUNTER__` -- looks like an MS extension? Is it widely supported?